tool/pub_get_offline.dart

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:io';
+
+final _client = HttpClient();
+
+/// Runs `dart pub get --offline --no-precompile` with pre-populated
+/// packages directly downloaded from the storage bucket.
+///
+/// This allows us to not depend on a running `pub.dev` while deploying
+/// a new version.
+Future<void> main(List<String> args) async {
+  final pkgDir = args.single;
+  final pubCacheDir = Directory('$pkgDir/.dart_tool/pub-cache').absolute;
+  print('PUB_CACHE: ${pubCacheDir.path}');
+  if (pubCacheDir.existsSync()) {
+    pubCacheDir.deleteSync(recursive: true);
+  }
+  pubCacheDir.createSync(recursive: true);
+
+  final pvs = _parsePubspecLockSync(File('$pkgDir/pubspec.lock'));
+  final packages = pvs.keys.toList()..sort();
+  for (final package in packages) {
+    final version = pvs[package]!;
+    print('Downloading $package: $version ...');
+    await _downloadInto(package, version, pubCacheDir);
```

Comment:
Maybe some simple retries will make us happy:
```suggestion
    var retry = 0;
    while (true) {
      print('Downloading $package: $version (attempt: ${retry+1})');
      try {
        await _downloadInto(package, version, pubCacheDir);
      } catch (e) {
        if (retry >= 8) {
          rethrow;
        }
        await Future.delayed(Duration(seconds: 1));
        retry++;
      }
    }
```